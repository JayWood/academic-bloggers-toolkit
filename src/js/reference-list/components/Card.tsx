import { observer } from 'mobx-react';
import * as React from 'react';

import { colors } from 'utils/styles';
import { createTooltip, destroyTooltip } from 'utils/Tooltips';

interface CardProps extends React.HTMLProps<HTMLDivElement> {
    readonly CSL: CSL.Data;
    readonly id: string;
    readonly index: string;
    readonly isSelected: boolean;
    readonly showTooltip: boolean;
    click(id: string, isSelected: boolean): void;
}

@observer
export class Card extends React.PureComponent<CardProps, {}> {
    timer: NodeJS.Timer;

    constructor(props: CardProps) {
        super(props);
    }

    click = () => {
        this.props.click(this.props.id, this.props.isSelected);
    };

    tooltip = (e: React.MouseEvent<HTMLDivElement>) => {
        if (!this.props.showTooltip) return;
        const t = e.currentTarget;
        this.timer = setTimeout(() => {
            createTooltip(t, this.props.index, 'left');
        }, 700) as any;
    };

    destroyTooltip = () => {
        if (!this.props.showTooltip) return;
        clearTimeout(this.timer);
        destroyTooltip();
    };

    render() {
        const { CSL, isSelected } = this.props;
        return (
            <div
                data-reference-id={this.props.id}
                role="menuitem"
                className={isSelected ? 'abt-card abt-card_selected' : 'abt-card'}
                onClick={this.click}
                onDoubleClick={this.props.onDoubleClick}
                onMouseEnter={this.tooltip}
                onMouseLeave={this.destroyTooltip}
            >
                <div>
                    {CSL.title}
                </div>
                <div className="abt-card__people">
                    {parsePeople(CSL.author)}
                </div>
                <div className="abt-card__meta-container">
                    <div className="abt-card__date">
                        ({parseDate(CSL.issued)})
                    </div>
                    <div className="abt-card__source">
                        {CSL.journalAbbreviation || CSL['container-title'] || CSL.publisher || ''}
                    </div>
                    <div className="abt-card__pages">
                        {CSL.page || 'n.p.'}
                    </div>
                </div>
                <style jsx>{`
                    .abt-card {
                        border-bottom: solid 1px ${colors.border};
                        padding: 5px 10px;
                        cursor: pointer;
                        font-weight: 300;
                        user-select: none;
                    }
                    .abt-card_selected {
                        border-left: solid #0073a1 3px;
                        padding: 5px 10px 5px 7.5px;
                    }
                    .abt-card__people {
                        font-size: 0.8em;
                        font-weight: 500;
                    }
                    .abt-card__meta-container {
                        display: flex;
                        font-size: 0.8em;
                        justify-content: space-between;
                        white-space: nowrap;
                    }
                    .abt-card__date {
                        padding-right: 3px;
                    }
                    .abt-card__source {
                        padding: 0 3px;
                        font-style: italic;
                        text-overflow: ellipsis;
                        overflow: hidden;
                    }
                    .abt-card__pages {
                        padding-left: 3px;
                    }
                `}</style>
            </div>
        );
    }
}

function parseDate(date?: CSL.Date | ''): string {
    if (!date) return 'n.d.';
    if (date.year) return `${date.year}`;
    if (
        date['date-parts'] &&
        date['date-parts']![0].length !== 0 &&
        date['date-parts']![0][0] !== undefined
    ) {
        return `${date['date-parts']![0][0]}`;
    }
    return 'n.d.';
}

function parsePeople(p?: CSL.Person[]): string {
    if (!p) return '';
    return p.slice(0, 4).reduce((prev, curr, i) => {
        if (i > 2) return prev;
        let name: string = '';
        if (curr.family && curr.given) {
            name = `${curr.family}, ${curr.given[0]}`;
        }
        if (curr.literal) {
            name = curr.literal;
        }
        if (name === '') return prev;
        switch (i) {
            case 0:
            case 1:
                return (prev += `${name}${p.length > i + 1 ? ', ' : '.'}`);
            case 2:
            default:
                return (prev += `${name}${p.length > i + 1 ? '...' : '.'}`);
        }
    }, '');
}
