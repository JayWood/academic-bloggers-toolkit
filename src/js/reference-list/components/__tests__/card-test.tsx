jest.useFakeTimers();
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import * as React from 'react';
import * as diff from 'snapshot-diff';
import Card from '../card';

const testData: { [id: string]: any } = {
    fourAuthors: {
        author: [
            {
                family: 'Smith',
                given: 'John',
            },
            {
                given: 'Bob',
                literal: 'Some literal name',
            },
            {
                family: 'Doe',
                given: 'Jane',
            },
            {
                family: 'Brown',
                given: 'Susan',
            },
        ],
        id: 'aaaaa',
        issued: {
            'date-parts': [['2016', '01', '01']],
        },
        title: 'Two valid authors',
        type: 'article',
    },
    noAuthorField: {
        id: 'aaaaa',
        title: 'No author field, No date field',
        type: 'article',
    },
    noAuthors: {
        author: [],
        id: 'aaaaa',
        issued: {
            'date-parts': [['2016', '01', '01']],
        },
        title: 'No authors',
        type: 'article',
    },
    singleAuthorInvalidField: {
        author: [
            {
                suffix: 'asdf',
            },
        ],
        id: 'aaaaa',
        issued: {
            year: 2010,
        },
        title: 'Single author with invalid field',
        type: 'article',
    },
    singleLiteralAuthorEdgeCase: {
        author: [
            {
                family: 'Smith',
                literal: 'I have no given name',
            },
        ],
        id: 'aaaaa',
        issued: {
            'date-parts': [[]],
        },
        title: 'Single author, literal name',
        type: 'article',
    },
    sixAuthors: {
        author: [
            {
                family: 'Smith',
                given: 'John',
            },
            {
                given: 'Bob',
                literal: 'Some literal name',
            },
            {
                family: 'Doe',
                given: 'Jane',
            },
            {
                family: 'Brown',
                given: 'Susan',
            },
            {
                family: 'Doe',
                given: 'Jane',
            },
            {
                family: 'Brown',
                given: 'Susan',
            },
        ],
        id: 'aaaaa',
        issued: {
            'date-parts': [['2016', '01', '01']],
        },
        title: 'Two valid authors',
        type: 'article',
    },
    threeAuthors: {
        author: [
            {
                family: 'Smith',
                given: 'John',
            },
            {
                given: 'Bob',
                literal: 'Some literal name',
            },
            {
                family: 'Doe',
                given: 'Jane',
            },
        ],
        id: 'aaaaa',
        issued: {
            'date-parts': [['2016', '01', '01']],
        },
        title: 'Two valid authors',
        type: 'article',
    },
    twoAuthors: {
        author: [
            {
                family: 'Smith',
                given: 'John',
            },
            {
                family: 'Doe',
                given: 'Jane',
            },
        ],
        id: 'aaaaa',
        issued: {
            'date-parts': [['2016', '01', '01']],
        },
        title: 'Two valid authors',
        type: 'article',
    },
};

const setup = (data: CSL.Data, selected: boolean = false, indexOnHover: boolean = true) => {
    const spy = jest.fn();
    const component = shallow(
        <Card
            isSelected={selected}
            CSL={data}
            onClick={spy}
            id={'id'}
            index={1}
            indexOnHover={indexOnHover}
        />,
    );
    return {
        component,
        date: component.find('.abt-card__date'),
        people: component.find('.abt-card__people'),
        spy,
    };
};

describe('<Card/>', () => {
    test('should render selected', () => {
        const { component } = setup(testData.noAuthors, true);
        expect(component.first().props().className).toBe('abt-card abt-card--selected');
    });
    test('should render unselected', () => {
        const { component } = setup(testData.noAuthors);
        expect(component.first().props().className).toBe('abt-card');
    });
    test('should call onClick when clicked', () => {
        const { component, spy } = setup(testData.noAuthors);
        component.simulate('click');
        expect(spy).toHaveBeenCalledTimes(1);
    });
    test('should render with no authors', () => {
        const { people } = setup(testData.noAuthors);
        expect(people.children().length).toBe(0);
    });
    test('should handle a situation where there are no authors, and no author field', () => {
        const { people } = setup(testData.noAuthorField);
        expect(people.children().length).toBe(0);
    });
    test('should handle a single author with invalid fields', () => {
        const { people } = setup(testData.singleAuthorInvalidField);
        expect(people.children().length).toBe(0);
    });
    test('should render with a single literal author', () => {
        const { people } = setup(testData.singleLiteralAuthorEdgeCase);
        expect(people.children().length).toBe(1);
        expect(people.text()).toBe('I have no given name.');
    });
    test('should format two authors correctly', () => {
        const { people } = setup(testData.twoAuthors);
        expect(people.children().length).toBe(1);
        expect(people.text()).toBe('Smith, J, Doe, J.');
    });
    test('should format three authors correctly', () => {
        const { people } = setup(testData.threeAuthors);
        expect(people.children().length).toBe(1);
        expect(people.text()).toBe('Smith, J, Some literal name, Doe, J.');
    });
    test('should format > 3 authors correctly', () => {
        const { people: people1 } = setup(testData.fourAuthors);
        expect(people1.children().length).toBe(1);
        expect(people1.text()).toBe('Smith, J, Some literal name, Doe, J...');
        const { people: people2 } = setup(testData.sixAuthors);
        expect(people2.children().length).toBe(1);
        expect(people2.text()).toBe('Smith, J, Some literal name, Doe, J...');
    });
    test('should parse an item without a date correctly', () => {
        const { date } = setup(testData.noAuthorField);
        expect(date.text()).toBe('(n.d.)');
    });
    test('should parse a "date-parts" date correctly', () => {
        const { date } = setup(testData.noAuthors);
        expect(date.text()).toBe('(2016)');
    });
    test('should parse a "year" date correctly', () => {
        const { date } = setup(testData.singleAuthorInvalidField);
        expect(date.text()).toBe('(2010)');
    });
    test('should handle strange edge-cases that result from upgrading from an older version', () => {
        const { date } = setup(testData.singleLiteralAuthorEdgeCase);
        expect(date.text()).toBe('(n.d.)');
    });
    test('should show and hide indices on hover when indexOnHover is set', () => {
        const { component } = setup(testData.fourAuthors, false, true);
        const initialRender = toJSON(component);
        const card = component.first();

        expect(initialRender).toMatchSnapshot();

        card.simulate('mouseenter');
        jest.runAllTimers();
        component.update();

        expect(diff(initialRender, toJSON(component))).toMatchSnapshot();

        card.simulate('mouseleave');
        expect(diff(initialRender, toJSON(component))).toMatchSnapshot();
    });
    test('should render appropriately with optional props unset', () => {
        const component = shallow(
            <Card
                isSelected={true}
                CSL={testData.fourAuthors}
                id="id"
                index={1}
                indexOnHover={false}
            />,
        );
        expect(toJSON(component)).toMatchSnapshot();
    });
});
