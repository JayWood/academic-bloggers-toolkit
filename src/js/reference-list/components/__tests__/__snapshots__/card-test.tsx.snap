// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card/> should render appropriately with optional props unset 1`] = `
<div
  className="abt-card abt-card--selected"
  id="aaaaa"
  role="menuitem"
>
  <div
    aria-hidden={true}
    className="item-number"
  >
    1
  </div>
  <div>
    Two valid authors
  </div>
  <div
    className="abt-card__people"
  >
    Smith, J, Some literal name, Doe, J...
  </div>
  <div
    className="abt-card__meta-container"
  >
    <div
      className="abt-card__date"
    >
      (2016)
    </div>
    <div
      className="abt-card__source"
    />
    <div
      className="abt-card__pages"
    >
      n.p.
    </div>
  </div>
  <style
    jsx={true}
  >
    
                        .abt-card {
                            position: relative;
                            border-bottom: solid 1px rgba(0, 0, 0, 0.12);
                            padding: 5px 10px;
                            cursor: pointer;
                            font-weight: 300;
                            user-select: none;
                        }
                        .abt-card--selected {
                            border-left: solid #0073a1 3px;
                            padding: 5px 10px 5px 7.5px;
                        }
                        .abt-card__people {
                            font-size: 0.8em;
                            font-weight: 500;
                        }
                        .abt-card__meta-container {
                            display: flex;
                            font-size: 0.8em;
                            justify-content: space-between;
                            white-space: nowrap;
                        }
                        .abt-card__date {
                            padding-right: 3px;
                        }
                        .abt-card__source {
                            padding: 0 3px;
                            font-style: italic;
                            text-overflow: ellipsis;
                            overflow: hidden;
                        }
                        .abt-card__pages {
                            padding-left: 3px;
                        }
                        .item-number {
                            position: absolute;
                            right: 0;
                            top: 50%;
                            visibility: hidden;
                            transform: translateY(-50%) translateX(100%);
                            padding: 5px 8px;
                            background: rgba(66, 66, 66, 0.7);
                            color: white;
                            border-top-left-radius: 2px;
                            border-bottom-left-radius: 2px;
                            font-weight: 500;
                        }
                        .item-number--active {
                            animation: slide-in 200ms cubic-bezier(0, 0, 0.2, 1) forwards;
                        }
                        @keyframes slide-in {
                            0% {
                                transform: translateY(-50%) translateX(100%);
                                opacity: 0;
                            }
                            50% {
                                transform: translateY(-50%) translateX(1%);
                            }
                            100% {
                                transform: translateY(-50%) translateX(0);
                                opacity: 1;
                                visibility: visible;
                            }
                        }
                    
  </style>
</div>
`;

exports[`<Card/> should show and hide indices on hover when indexOnHover is set 1`] = `
<div
  className="abt-card"
  id="aaaaa"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  role="menuitem"
>
  <div
    aria-hidden={true}
    className="item-number"
  >
    1
  </div>
  <div>
    Two valid authors
  </div>
  <div
    className="abt-card__people"
  >
    Smith, J, Some literal name, Doe, J...
  </div>
  <div
    className="abt-card__meta-container"
  >
    <div
      className="abt-card__date"
    >
      (2016)
    </div>
    <div
      className="abt-card__source"
    />
    <div
      className="abt-card__pages"
    >
      n.p.
    </div>
  </div>
  <style
    jsx={true}
  >
    
                        .abt-card {
                            position: relative;
                            border-bottom: solid 1px rgba(0, 0, 0, 0.12);
                            padding: 5px 10px;
                            cursor: pointer;
                            font-weight: 300;
                            user-select: none;
                        }
                        .abt-card--selected {
                            border-left: solid #0073a1 3px;
                            padding: 5px 10px 5px 7.5px;
                        }
                        .abt-card__people {
                            font-size: 0.8em;
                            font-weight: 500;
                        }
                        .abt-card__meta-container {
                            display: flex;
                            font-size: 0.8em;
                            justify-content: space-between;
                            white-space: nowrap;
                        }
                        .abt-card__date {
                            padding-right: 3px;
                        }
                        .abt-card__source {
                            padding: 0 3px;
                            font-style: italic;
                            text-overflow: ellipsis;
                            overflow: hidden;
                        }
                        .abt-card__pages {
                            padding-left: 3px;
                        }
                        .item-number {
                            position: absolute;
                            right: 0;
                            top: 50%;
                            visibility: hidden;
                            transform: translateY(-50%) translateX(100%);
                            padding: 5px 8px;
                            background: rgba(66, 66, 66, 0.7);
                            color: white;
                            border-top-left-radius: 2px;
                            border-bottom-left-radius: 2px;
                            font-weight: 500;
                        }
                        .item-number--active {
                            animation: slide-in 200ms cubic-bezier(0, 0, 0.2, 1) forwards;
                        }
                        @keyframes slide-in {
                            0% {
                                transform: translateY(-50%) translateX(100%);
                                opacity: 0;
                            }
                            50% {
                                transform: translateY(-50%) translateX(1%);
                            }
                            100% {
                                transform: translateY(-50%) translateX(0);
                                opacity: 1;
                                visibility: visible;
                            }
                        }
                    
  </style>
</div>
`;

exports[`<Card/> should show and hide indices on hover when indexOnHover is set 2`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -5,12 +5,12 @@
   onMouseEnter={[Function res]}
   onMouseLeave={[Function res]}
   role=\\"menuitem\\"
 >
   <div
-    aria-hidden={true}
-    className=\\"item-number\\"
+    aria-hidden={false}
+    className=\\"item-number item-number--active\\"
   >
     1
   </div>
   <div>
     Two valid authors"
`;

exports[`<Card/> should show and hide indices on hover when indexOnHover is set 3`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;
